base/client.jl

```diff
@@ -548,13 +538,28 @@ function _start()
     append!(ARGS, Core.ARGS)
     # clear any postoutput hooks that were saved in the sysimage
     empty!(Base.postoutput_hooks)
+    local ret = 0
     try
-        exec_options(JLOptions())
+        should_run_repl = exec_options(JLOptions())
+        if should_run_repl
+            if isassigned(REPL_MODULE_REF)
+                ret = REPL_MODULE_REF[].main(ARGS)
+            end
+        elseif isdefined(Main, :main)
```

Comment:
Testing `isdefined` will cause the `Main.main` binding to become resolved (in the precompile process, if exported by a package), which may cause problems for people who want to add their own `Main.main` later? Binding resolution time is one of the less well-defined semantics unfortunately for this. Maybe we should check `isbindingresolved`?

Reply:
I do want to resolve the inputs for actually executing the thing, but I can see the case for allowing the binding to be unresolved during precompile. I'll tweak this.

Reply:
I've changed my mind on this - I do think you want to resolve the binding in the precompile process, so you get your warning if there's conflicts and the `-e using` pattern works without surprises. If you don't want a package's `main` function, just import it.

Reply:
You don't even need to `import` the package - if there's already an existing `main`, that one will still be resolved correctly. As far as I can tell, the conflict just doesn't make the modules `main` available at all:

```
[sukera@tower ~]$ julia -q
julia> foo() = 4
foo (generic function with 1 method)

julia> module A
           export foo
           foo() = 1
       end
Main.A

julia> using .A
WARNING: using A.foo in module Main conflicts with an existing identifier.

julia> foo()
4
```

The message could be tweaked/special cased for `main`, but that can be done later together with improving the warning for two packages trying to bring a `main` into the namespace.

Reply:
It is probably a moot point anyways. I was noticing that the export main pattern actually may be going to be discouraged, looking at the example here of REPL, so that explicit import of the intended module is needed anyways. And incremental compile packages are not permitted to change the set of usings for Main anyways, so they shouldn't add a binding there. So seems fine.